src/core/sections.js

```diff
@@ -0,0 +1,118 @@
+/**
+ * Module core/sections
+ *
+ * Adds sections to the document, based on the heading structure.
+ */
+export const name = "core/sections";
+
+class Builder {
+  constructor(doc) {
+    this.doc = doc;
+    this.root = doc.createDocumentFragment();
+    this.stack = [this.root];
+    this.current = this.root;
+  }
+  findPosition(header) {
+    return parseInt(header.tagName.charAt(1), 10);
+  }
+  findParent(position) {
+    let parent;
+    while (position > 0) {
+      position--;
+      parent = this.stack[position];
+      if (parent) return parent;
+    }
+  }
+  findHeader({ firstChild: node }) {
+    while (node) {
+      if (/H[1-6]/.test(node.tagName)) {
+        return node;
+      }
+      node = node.nextSibling;
+    }
+    return null;
+  }
+
+  addHeader(header) {
+    const section = this.doc.createElement("section");
+    const position = this.findPosition(header);
+
+    section.appendChild(header);
+    this.findParent(position).appendChild(section);
+    this.stack[position] = section;
+    this.stack.length = position + 1;
+    this.current = section;
+  }
+
+  addSection(node, process) {
+    const header = this.findHeader(node);
+    const position = header ? this.findPosition(header) : 1;
+    const parent = this.findParent(position);
+
+    if (header) {
+      node.removeChild(header);
+    }
+
+    node.appendChild(process(node));
+
+    if (header) {
+      node.prepend(header);
+    }
+
+    parent.appendChild(node);
+    this.current = parent;
+  }
+
+  addElement(node) {
+    this.current.appendChild(node);
+  }
+}
+
+function structure(fragment, doc) {
+  function process(root) {
+    const stack = new Builder(doc);
+    while (root.firstChild) {
+      const node = root.firstChild;
+      switch (node.localName) {
+        case "h1":
+        case "h2":
+        case "h3":
+        case "h4":
+        case "h5":
+        case "h6":
+          stack.addHeader(node);
+          break;
+        case "section":
+          stack.addSection(node, process);
+          break;
+        default:
+          stack.addElement(node);
+      }
+    }
+    return stack.root;
+  }
+  return process(fragment);
+}
+
+/**
+ * Re-structure DOM around elem whose markdown has been processed.
+ * @param {Element} elem
+ */
+export function restructure(elem) {
+  const structuredInternals = structure(elem, elem.ownerDocument);
+  if (
+    structuredInternals.firstElementChild.localName === "section" &&
+    elem.localName === "section"
+  ) {
+    const section = structuredInternals.firstElementChild;
+    section.remove();
+    elem.append(...section.childNodes);
+  } else {
+    elem.textContent = "";
+  }
+  elem.appendChild(structuredInternals);
+}
+
+export function run() {
+  restructure(document.body);
```

Comment:
Should we make it configurable? Like, run if "markdown" and "????".... or, just make it be the standard operating mode?

Reply:
I think making it default should be fine, as long as resultant output is same.